org.eclipse.nebula.widgets.nattable.core/src/org/eclipse/nebula/widgets/nattable/filterrow/combobox/FilterRowComboBoxCellEditor.java

```diff
@@ -264,7 +264,7 @@ && getComboBoxDataProvider() instanceof FilterRowComboBoxDataProvider
             // available items
             List<?> allValues = ((FilterRowComboBoxDataProvider) getComboBoxDataProvider()).getAllValues(getColumnIndex());
             List<?> visibleValues = getComboBoxDataProvider().getValues(getColumnIndex(), getRowIndex());
-            List<?> diffValues = new ArrayList<>(allValues);
+            HashSet<?> diffValues = new HashSet<>(allValues);
             diffValues.removeAll(visibleValues);
```

Comment:
I just realized: This does not actually solve the issue:   
The expensive `contains` operation within the `removeAll` is not executed on "this" (ie. `diffValues`) but the method argument, an arraylist in this case. See `HashSet`, l. 175 (Java 11).  
Sorry, my bad.

Reply:
Before I changed from `ArrayList` to `HashSet` I tested locally with this code:

```java
    @Test
    void testListPerformance() {
        ArrayList<String> all = new ArrayList<>();
        ArrayList<String> flanders = new ArrayList<>();

        for (int i = 0; i < 100_000; i++) {
            all.add("Simpson_" + i);
            all.add("Flanders_" + i);
            flanders.add("Flanders_" + i);
        }

        long start = System.currentTimeMillis();

        ArrayList<String> diff = new ArrayList<>(all);
        diff.removeAll(flanders);

        long end = System.currentTimeMillis();

        System.out.println("ArrayList#removeAll(ArrayList) - " + (end - start));

        start = System.currentTimeMillis();

        HashSet<String> diffSet = new HashSet<>(all);
        diffSet.removeAll(flanders);

        end = System.currentTimeMillis();

        System.out.println("HashSet#removeAll(ArrayList) - " + (end - start));

        start = System.currentTimeMillis();

        HashSet<String> diffSetSet = new HashSet<>(all);
        diffSetSet.removeAll(new HashSet<>(flanders));

        end = System.currentTimeMillis();

        System.out.println("HashSet#removeAll(HashSet) - " + (end - start));
    }
```

If `diff` is a List, the operation takes about 60s. The `diffSet` variant takes around 30ms. I would say that the test is similar to the code in discussion. Or what am I missing?

Reply:
OK, I found out why my test was working. It strongly depends on the size of the collection passed as parameter. As long as the collection is smaller than the "this" collection, things work fast. If the collection is bigger, the performance is a mess.

I adjusted the collection creation to verify it:

```java
        for (int i = 0; i < 100_000; i++) {
            all.add("Simpson_" + i);
            all.add("Flanders_" + i);
            flanders.add("Flanders_" + i);
            flanders.add("FlandersO_" + i);
        }

        for (int i = 0; i < 10_000; i++) {
            flanders.add("FlandersXxx_" + i);
        }
```

Now the _HashSet#removeAll(ArrayList)_ takes 123s while _HashSet#removeAll(HashSet)_ is still at about 30ms.